File: src/types/project.ts
import type { ThemeConfig } from './theme'

export const PROJECT_STATUS_VALUES = [
  'draft',
  'planning',
  'generating',
  'editing',
  'completed',
  'archived'
] as const

export type ProjectStatus = (typeof PROJECT_STATUS_VALUES)[number]

export const PAGE_STATUS_VALUES = ['planned', 'generating', 'generated', 'optimizing', 'confirmed'] as const

export type PageStatus = (typeof PAGE_STATUS_VALUES)[number]

export interface OutlinePage {
  content_brief: string
  data_refs: string[]
  layout: string
  page_number: number
  title: string
  type: string
}

export interface Outline {
  pages: OutlinePage[]
  theme_suggestion: string
  title: string
  total_pages: number
}

export interface ProjectPage {
  chat_message_count: number
  created_at: string
  id: string
  page_number: number
  page_type: string | null
  project_id: string
  status: PageStatus
  title: string | null
  updated_at: string
  version: number
  vue_code: string | null
}

export interface PageVersion {
  change_description: string | null
  created_at: string
  id: string
  page_id: string
  version: number
  vue_code: string
}

export interface PageInsertAfterPayload {
  description: string
}

export interface PageMutationResponse {
  success: boolean
}

export interface PageReorderPayload {
  page_numbers: number[]
}

export interface Project {
  created_at: string
  description: string | null
  id: string
  name: string
  status: ProjectStatus
  theme_config: ThemeConfig | null
  total_pages: number
  updated_at: string
}

export interface ProjectCreatePayload {
  description?: string | null
  name: string
}

export interface ProjectUpdatePayload {
  description?: string | null
  name?: string
  status?: ProjectStatus
}

export interface ProjectListResponse {
  projects: Project[]
  total: number
}

export interface ProjectDetailResponse extends Project {
  outline: Outline | null
  pages: ProjectPage[]
}

export interface ProjectDeleteResponse {
  success: boolean
}

File: python-backend/app/schemas/project.py
from __future__ import annotations

from datetime import datetime
from pydantic import Field, field_validator

from app.schemas.base import APIModel, ORMModel
from app.schemas.enums import PageStatus, ProjectStatus
from app.schemas.theme import ThemeConfig


class OutlinePageSchema(APIModel):
    page_number: int
    title: str
    type: str
    content_brief: str
    layout: str
    data_refs: list[str] = Field(default_factory=list)


class OutlineSchema(APIModel):
    title: str
    total_pages: int
    theme_suggestion: str
    pages: list[OutlinePageSchema] = Field(default_factory=list)


class ProjectCreate(APIModel):
    name: str = Field(min_length=1, max_length=255)
    description: str | None = None

    @field_validator("name")
    @classmethod
    def validate_name(cls, value: str) -> str:
        normalized = value.strip()
        if not normalized:
            raise ValueError("Project name cannot be blank.")
        return normalized

    @field_validator("description")
    @classmethod
    def normalize_description(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = value.strip()
        return normalized or None


class ProjectUpdate(APIModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = None
    status: ProjectStatus | None = None

    @field_validator("name")
    @classmethod
    def validate_optional_name(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = value.strip()
        if not normalized:
            raise ValueError("Project name cannot be blank.")
        return normalized

    @field_validator("description")
    @classmethod
    def normalize_optional_description(cls, value: str | None) -> str | None:
        if value is None:
            return None
        normalized = value.strip()
        return normalized or None


class ProjectResponse(ORMModel):
    id: str
    name: str
    description: str | None = None
    status: ProjectStatus
    theme_config: ThemeConfig | None = None
    total_pages: int
    created_at: datetime
    updated_at: datetime


class PageResponse(ORMModel):
    id: str
    project_id: str
    page_number: int
    title: str | None = None
    page_type: str | None = None
    vue_code: str | None = None
    status: PageStatus
    version: int
    created_at: datetime
    updated_at: datetime


class PageVersionResponse(ORMModel):
    id: str
    page_id: str
    version: int
    vue_code: str
    change_description: str | None = None
    created_at: datetime


class ProjectDetailResponse(ProjectResponse):
    outline: OutlineSchema | None = None
    pages: list[PageResponse] = Field(default_factory=list)


class ProjectListResponse(APIModel):
    projects: list[ProjectResponse] = Field(default_factory=list)
    total: int


class ProjectDeleteResponse(APIModel):
    success: bool = True
